File: packages/core/src/ingest.ts
import type { AppSettings } from "@localrag/config";
import type { Db } from "@localrag/db";
import {
  findDocumentByHash,
  insertDocument,
  logIngestionEvent,
  replaceDocumentChunksPool,
} from "@localrag/db";
import { chunkText, resolveParser, sha256Hex } from "@localrag/ingestion";
import { embedText } from "@localrag/llm";
import type pg from "pg";

export type IngestFileInput = {
  buffer: Buffer;
  fileName: string;
  sourcePath?: string | null;
};

export type IngestFileResult =
  | { ok: true; documentId: string; deduped: boolean }
  | { ok: false; error: string };

function assertEmbeddingDim(vec: number[], expected: number) {
  if (vec.length !== expected) {
    throw new Error(
      `Embedding dimension mismatch: got ${vec.length}, expected ${expected}`,
    );
  }
}

export async function ingestFile(params: {
  db: Db;
  pool: pg.Pool;
  settings: AppSettings;
  apiKey: string | null;
  input: IngestFileInput;
}): Promise<IngestFileResult> {
  const { db, pool, settings, apiKey, input } = params;
  const started = Date.now();
  let documentId: string | null = null;
  try {
    const parser = resolveParser(input.fileName);
    if (!parser) {
      return { ok: false, error: "Unsupported file type" };
    }

    const parsed = await parser.parse({
      buffer: input.buffer,
      fileName: input.fileName,
    });
    const hash = sha256Hex(parsed.text);
    const existing = await findDocumentByHash(db, hash);
    if (existing) {
      return { ok: true, documentId: existing.id, deduped: true };
    }

    const doc = await insertDocument(db, {
      sourcePath: input.sourcePath ?? null,
      title: parsed.title,
      mime: parsed.mime,
      contentHash: hash,
      status: "ready",
    });
    documentId = doc.id;

    const parts = chunkText(
      parsed.text,
      settings.rag.chunkSize,
      settings.rag.chunkOverlap,
    );

    const rows = [];
    for (const p of parts) {
      const emb = await embedText(settings, apiKey, p.content);
      assertEmbeddingDim(emb, settings.rag.embeddingDimensions);
      rows.push({
        documentId: doc.id,
        chunkIndex: p.index,
        content: p.content,
        charStart: p.charStart,
        charEnd: p.charEnd,
        embedding: emb,
      });
    }

    await replaceDocumentChunksPool(pool, doc.id, rows);

    await logIngestionEvent(db, {
      documentId: doc.id,
      stage: "complete",
      bytes: input.buffer.byteLength,
      durationMs: Date.now() - started,
    });

    return { ok: true, documentId: doc.id, deduped: false };
  } catch (e) {
    const message = e instanceof Error ? e.message : String(e);
    await logIngestionEvent(db, {
      documentId,
      stage: "error",
      error: message,
      bytes: input.buffer.byteLength,
      durationMs: Date.now() - started,
    });
    return { ok: false, error: message };
  }
}

File: rag-service/rag_service/services/ingest.py
from __future__ import annotations

import time
from typing import Any

from rag_service.contracts import AppSettingsPayload
from rag_service.db.pool import get_pool
from rag_service.db import repositories as repo
from rag_service.domain.chunking import chunk_text
from rag_service.domain.hash import sha256_hex
from rag_service.llm import embeddings as emb
from rag_service.parsers import parse_file


def ingest_file(
    *,
    buffer: bytes,
    file_name: str,
    source_path: str | None,
    settings: AppSettingsPayload,
    api_key: str | None,
) -> dict[str, Any]:
    pool = get_pool()
    started = time.time() * 1000
    started_ms = int(started)
    document_id: str | None = None
    try:
        parsed = parse_file(file_name, buffer)
        if parsed is None:
            return {"ok": False, "error": "Unsupported file type"}

        h = sha256_hex(parsed.text)
        existing = repo.find_document_by_hash(pool, h)
        if existing:
            return {"ok": True, "documentId": existing, "deduped": True}

        doc_id = repo.insert_document(
            pool,
            source_path=source_path,
            title=parsed.title,
            mime=parsed.mime,
            content_hash=h,
            status="ready",
        )
        document_id = doc_id

        parts = chunk_text(
            parsed.text,
            settings.rag.chunk_size,
            settings.rag.chunk_overlap,
        )
        rows: list[dict[str, Any]] = []
        for p in parts:
            vec = emb.embed_text(settings, api_key, p.content)
            if len(vec) != settings.rag.embedding_dimensions:
                raise ValueError(
                    f"Embedding dimension mismatch: got {len(vec)}, expected {settings.rag.embedding_dimensions}"
                )
            rows.append(
                {
                    "chunkIndex": p.index,
                    "content": p.content,
                    "charStart": p.char_start,
                    "charEnd": p.char_end,
                    "embedding": vec,
                }
            )

        repo.replace_document_chunks(pool, doc_id, rows)

        repo.log_ingestion_event(
            pool,
            document_id=doc_id,
            stage="complete",
            error=None,
            bytes_=len(buffer),
            duration_ms=int(time.time() * 1000) - started_ms,
        )
        return {"ok": True, "documentId": doc_id, "deduped": False}
    except Exception as e:
        msg = str(e)
        repo.log_ingestion_event(
            pool,
            document_id=document_id,
            stage="error",
            error=msg,
            bytes_=len(buffer),
            duration_ms=int(time.time() * 1000) - started_ms,
        )
        return {"ok": False, "error": msg}
